Add autowiring-support for PHP7.4 typed properties

bzikarsky opened the pull request:
The current implementation prioritizes `@var` and the `name` property of the `@Inject` annotation over the type of the property. This allows to still inject custom-named entries into the properties even in the presence of types.

Fixes #707

bzikarsky commented:
I get a bunch of formatting errors on source files I did not touch, is this intentional?

bzikarsky commented:
It looks like, phpstan needs to be pushed to PHP7.4 only - I am not sure if you want me to do this in this branch? If you update master, I'll happily rebase.

mnapoli commented:
That pull request is awesome! Thank you for covering all cases!

I will try to sort out master, hang on ^^

mnapoli commented:
👍  I updated master, you can rebase/merge it in your branch.

bzikarsky commented:
phpstan pointed to me to two issues, I now fixed (and squashed):

1)  `getName()` only exists for `ReflectionNamedType`

2) I didn't handle scalar-typed properties "correctly" (the same way `@var` does)
    This turned out to be a bit more complex and is not 100% fixed. The separate annotation-reader project has an internal list of ignored types which I couldn't access but also didn't want to replicate. So I decided that I ignore the type on the typed property if I cannot locate a class or interface with the name. It's probably close enough to the desired behaviour. WDYT?

mnapoli commented on src/Definition/Source/AnnotationBasedAutowiring.php:
```diff
@@ -125,9 +126,18 @@ private function readProperty(ReflectionProperty $property, ObjectDefinition $de
+        // Try using PHP7.4 typed properties
+        if (\PHP_VERSION_ID > 70400
+            && $entryName === null
+            &&  $property->getType() instanceof ReflectionNamedType
```
```suggestion
            && $property->getType() instanceof ReflectionNamedType
```

mnapoli commented:
Yes that looks good to me.

There is one last thing to fix regarding code formatting, I opened a code suggestion.

bzikarsky commented:
Thanks, missed this. PR is updated with the squashed change.

bzikarsky commented:
Do we have an ETA on this being merged (and tagged incl. the other PR)? :)

bzikarsky commented:
Ping :) -- I have a bunch of developers eagerly waiting for this feature and I'd rather not switch our projects to our forked repo.

mnapoli commented:
Sorry, I disable email notifications because I get so many of them every day, it creates a lot of anxiety :)

Let's give that a go!

bzikarsky commented:
Awesome, thanks. I'll give it a go in prod today. :)
